Add support for pointers to reflect.Kind types

pandada8 opened the pull request:
this should resolve GODRIVER-535, pointers handling is added to both fast path and reflection path.

skriptble commented on bson/bsoncodec/default_value_decoders.go:
```diff
@@ -62,33 +62,66 @@ func (dvd DefaultValueDecoders) RegisterDefaultDecoders(rb *RegistryBuilder) {
+	}
+
+	if val.Kind() == reflect.Ptr {
+		val.Set(reflect.New(val.Type().Elem()))
```
We should only allocate a new value here if the val is `nil`.

pandada8 replied:
a new function `getValue` is introduced to check and allocate new value.  
do you have any idea of a better name ?

skriptble commented on bson/bsoncodec/default_value_decoders.go:
```diff
@@ -62,33 +62,66 @@ func (dvd DefaultValueDecoders) RegisterDefaultDecoders(rb *RegistryBuilder) {
-	if val.Type().Kind() != reflect.Bool {
+
+	if notKindAndPtrToKind(val, reflect.Bool) {
 		return ValueDecoderError{Name: "BooleanDecodeValue", Types: []interface{}{bool(true)}, Received: i}
```
Types should be `*bool` and `**bool`.

pandada8 replied:
all related types should be updated

skriptble commented on bson/bsoncodec/default_value_decoders.go:
```diff
@@ -197,7 +277,12 @@ func (dvd DefaultValueDecoders) IntDecodeValue(dc DecodeContext, vr bsonrw.Value
 
-	val.SetInt(i64)
+	if val.Type().Kind() == reflect.Ptr {
+		val.Set(reflect.New(val.Type().Elem()))
```
We should also only alloc if val is `nil`.

skriptble commented on bson/bsoncodec/default_value_decoders.go:
```diff
@@ -452,17 +627,22 @@ func (dvd DefaultValueDecoders) Decimal128DecodeValue(dctx DecodeContext, vr bso
-	if !ok || target == nil {
+	switch target := i.(type) {
+	case *decimal.Decimal128:
+		d128, err := vr.ReadDecimal128()
```
Is there a reason you moved the `ReadDecimal128` method invocation here? It adds repetitive code and extra errors to check.

pandada8 replied:
changed EmptyInterfaceDecodeValue's closure style

skriptble commented on bson/bsoncodec/struct_codec.go:
```diff
@@ -97,7 +97,13 @@ func (sc *StructCodec) EncodeValue(r EncodeContext, vw bsonrw.ValueWriter, i int
 		ectx := EncodeContext{Registry: r.Registry, MinSize: desc.minSize}
-		err = encoder.EncodeValue(ectx, vw2, rv.Interface())
+
+		// HACK: pass *Document as is
```
Why is this here? The `bson.Document` decoder should only be handling `**bson.Document`.

pandada8 commented:
still finding the elegant checking process for most typing. I may push a new version in weekend

delioda79 commented on bson/bsoncodec/struct_codec.go:
```diff
@@ -86,8 +86,7 @@ func (sc *StructCodec) EncodeValue(r EncodeContext, vw bsonrw.ValueWriter, i int
 		}
-
-		if desc.omitEmpty && iszero(rv.Interface()) {
+		if (rv.Kind() == reflect.Ptr || desc.omitEmpty) && iszero(rv.Interface()) {
```
Hi, I was testing this but I think this line is wrong. If we do not have omitempty and the kind is pointer we want to add the key just with a null value in case the pointer is a nil pointer. Problem here is that iszero does not work with pointers and apparently is defined only for objectid

skriptble replied:
That is correct, we want to write a BSON Null if the pointer is nil, not skip it entirely.

delioda79 replied:
So yes this code here is wrong as it is skipping any nil pointer

pandada8 replied:
I update the code, now `WriteNull` is called when an nil pointer is encountered.
Should I put it in a standalone function ?

skriptble replied:
For the encode path, we should just let the individual encoders handle it, we don't need to have the struct codec care if it has a `nil` pointer. We should add `nil` pointer handling for all of the  encoders as well as the decoders. This will allow users to create an encoder that handles `nil` pointers differently, e.g. if a user wants to put some default data instead of BSON null.

pandada8 replied:
You're right, but I have no idea how to RegisterEncoder for a nil value in the current registry. Maybe we should modify these encoders to add pointer support?

skriptble commented on bson/bsoncodec/default_value_decoders.go:
```diff
@@ -62,15 +63,88 @@ func (dvd DefaultValueDecoders) RegisterDefaultDecoders(rb *RegistryBuilder) {
 func (dvd DefaultValueDecoders) BooleanDecodeValue(dctx DecodeContext, vr bsonrw.ValueReader, i interface{}) error {
-	if vr.Type() != bsontype.Boolean {
-		return fmt.Errorf("cannot decode %v into a boolean", vr.Type())
+	var (
```
Instead of using a multiline `var` block, can you please change this to individual `var` statements?

pandada8 replied:
all var block in default_value_decoders is changed

skriptble commented on bson/bsoncodec/default_value_decoders.go:
```diff
@@ -62,15 +63,88 @@ func (dvd DefaultValueDecoders) RegisterDefaultDecoders(rb *RegistryBuilder) {
+		b      bool
+		isNull bool
+	)
+	readValue := func() {
```
I think we're adding a little too much indirection here. Let's instead just split off the fast path decoders into separate methods, and call them from this function if the type matches. It's fine to duplicate this specific logic between the methods.

pandada8 replied:
I am not sure I understood correctly, so I only update the Boolean case. If you think it's ok, I can work on the rest functions

skriptble replied:
I added a patch with an updated [`BooleanDecodeValue`](https://github.com/mongodb/mongo-go-driver/commit/dc6a94e7fc31b407ad13fbd3d2e11a6ff069c76d#diff-d7870a7e54c1393f5080c1fbf611666dR114), can you update your patch to follow this design? If you want to do some things to make the code more compact, feel free.

Trial97 commented:
Hello guys,

How it's going with this PR?

Thanks,
Trial97

pandada8 commented:
It's been a long and tough week for me, moved into a new apartment and had a fever, hope it's not too late to submit some updated patch.

I rebased my branch to resolve some merge conflict and add new error type. more update on Decoder function should come in tomorrow.

skriptble commented:
Hi @pandada8,

We've found a different approach that requires less code overall and makes the code easier to maintain, so I'm going to close this pull request. Thank you for all your hard work on this pull request, and I'm sorry that we cannot incorporate these changes into the BSON library at this time. Please do contribute in the future! We always welcome contributions from the community.

If you want to follow along with the progress of the new design, you can do so [here](https://review.gerrithub.io/c/mongodb/mongo-go-driver/+/435374).

Thanks again for your hard work!
--Kris
